pkg/kv/kvserver/replica_application_cmd.go

```diff
@@ -34,6 +35,86 @@ import (
 //
 // These allow Replica to interface with the storage/apply package.
 
+// replicatedCmdBase is the part of a replicatedCmd relevant for stand-alone log
+// application, i.e. without having a surrounding *Replica present. In that
+// case, it is assumed that only one replica is processed at a given time[^1], no
+// in-memory state is kept, and no clients are waiting for responses to the
+// commands. This obviates the need for concurrency control, in-memory state
+// updates, reproposals, etc.
+//
+// [^1]: at least when the current command includes a split or a merge, i.e.
+// changes replica boundaries.
+type replicatedCmdBase struct {
+	*raftlog.Entry
+	// The following fields are set in shouldApplyCommand when we validate that
+	// a command applies given the current lease and GC threshold. The process
+	// of setting these fields is what transforms an apply.Command into an
+	// apply.CheckedCommand.
+	leaseIndex    uint64
+	forcedErr     *roachpb.Error
+	proposalRetry kvserverbase.ProposalRejectionType
+}
+
+var _ apply.Command = (*replicatedCmdBase)(nil)
+var _ apply.CheckedCommand = (*replicatedCmdBase)(nil)
+var _ apply.AppliedCommand = (*replicatedCmdBase)(nil)
+
+// decode populates the receiver from the provided entry.
+func (c *replicatedCmdBase) decode(e *raftpb.Entry) error {
+	if c.Entry != nil {
+		c.Entry.Release()
+		c.Entry = nil
+	}
+
+	var err error
+	c.Entry, err = raftlog.NewEntry(*e)
+	if err != nil {
+		return wrapWithNonDeterministicFailure(err, "while decoding raft entry")
+	}
+	return nil
+}
+
+// Index implements apply.Command.
+func (c *replicatedCmdBase) Index() uint64 {
+	return c.Entry.Index
+}
+
+// IsTrivial implements apply.Command.
+func (c *replicatedCmdBase) IsTrivial() bool {
+	return isTrivial(c.replicatedResult())
+}
+
+// IsLocal implements apply.Command.
+func (c *replicatedCmdBase) IsLocal() bool {
+	return false
+}
+
+// Ctx implements apply.Command.
+func (c *replicatedCmdBase) Ctx() context.Context {
+	return context.Background()
+}
+
+// AckErrAndFinish implements apply.Command.
+func (c *replicatedCmdBase) AckErrAndFinish(context.Context, error) error {
+	return nil
+}
+
+// AckOutcomeAndFinish implements apply.AppliedCommand.
+func (c *replicatedCmdBase) AckOutcomeAndFinish(context.Context) error { return nil }
+
+// Rejected implements apply.CheckedCommand.
+func (c *replicatedCmdBase) Rejected() bool { return c.forcedErr != nil }
+
+// CanAckBeforeApplication implements apply.CheckedCommand.
+func (c *replicatedCmdBase) CanAckBeforeApplication() bool { return false }
+
+// AckSuccess implements apply.CheckedCommand.
+func (c *replicatedCmdBase) AckSuccess(context.Context) error { return nil }
+
+func (c *replicatedCmdBase) replicatedResult() *kvserverpb.ReplicatedEvalResult {
+	return &c.Entry.Cmd.ReplicatedEvalResult
+}
```

Comment:
> Some of these are duplicated in replicatedCmd. Does replicatedCmdBase need to implement the full interface? Will "base" ever be used with all these methods? If not, could there be a sub-interface that excludes unused stuff?

Yes, it will implement the entire interface, see the PR description - `ReplicatedCmdBase` will be the `apply.Command` used for standalone application.
I'm not planning on picking a fight with the `apply` interfaces since I want to focus on the refactors that are required for standalone application, but there could probably be sub-interfaces, where standalone application operates only on the sub-interfaces.

> Not sure this is related, but I think in general interfaces in Go are made to workaround types, rather than types workaround interfaces. Often types aren't even aware of interfaces. For the same reason, I usually prefer comments to describe what the type does, instead of what it implements (unless the interface is an intended part of the API).

Types don't know about their surrounding interface, but the interface provides the contract - the caller doesn't know the concrete type and thus can't be expected to know how to use it.
There is often some ambiguity but whenever an implementation does something that's not obvious from the interface I think it is the *interface* that needs to reflect that. Whether the method comment spells it our or not is sort of secondary.
But, I agree with what I think your concrete point is, which is that some of these methods could use an additional comment that explains what they're doing and how that relates to the interface. Will add that.

I'll also spend a few minutes on the subinterface. Can probably group a few methods together and make it easier to see what's going on that way.

Reply:
> I'll also spend a few minutes on the subinterface. Can probably group a few methods together and make it easier to see what's going on that way.

Decided not to do anything at this point, since there are three `Command` interfaces and while we could extract out a subinterface on each of them, that would cause more confusion than benefit. It would only make sense if we could also only implement the base interfaces, but that then requires changes to the apply machinery, which I want to avoid as they may not even be helpful. For example, we'd have to type-assert on the "wider" interfaces on each command application, which pessimizes the production code. I think this is better as is but either way now's not a good time to pull on this particular thread.

Reply:
Improved commentary though.